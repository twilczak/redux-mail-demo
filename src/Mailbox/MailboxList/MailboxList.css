.mailbox-list {
  margin: 0;
  padding: 0 10px;
  list-style-type: none;
  list-style-position: inside;
  height: 205px;
  overflow-y: auto; }

.mailbox-list-item,
.mailbox-list-header {
  display: block;
  padding: 0 5px; }

.mailbox-list-header {
  font-weight: 600; }

.mailbox-list-item {
  transition: background .25s; }
  .mailbox-list-item:last-child {
    margin-bottom: 10px; }
  .mailbox-list-item:hover {
    background: rgba(225, 225, 225, 0.5); }
  .mailbox-list-item.is-selected {
    color: dodgerblue; }

.mailbox-item-link {
  display: block;
  text-decoration: none;
  cursor: pointer; }
  .mailbox-item-link.is-active {
    color: dodgerblue; }

.mailbox-list-item .date-sent,
.mailbox-list-item .recipient,
.mailbox-list-item .sender,
.mailbox-list-item .subject,
.mailbox-list-header .date-sent,
.mailbox-list-header .recipient,
.mailbox-list-header .sender,
.mailbox-list-header .subject {
  display: inline-block;
  padding: 5px; }

.mailbox-list-item .date-sent,
.mailbox-list-header .date-sent {
  width: 15%; }

.mailbox-list-item .recipient,
.mailbox-list-item .sender,
.mailbox-list-header .recipient,
.mailbox-list-header .sender {
  border-right: 1px solid black;
  width: 24%; }

.mailbox-list-item .subject,
.mailbox-list-header .subject {
  width: 50%;
  border-right: 1px solid black; }
