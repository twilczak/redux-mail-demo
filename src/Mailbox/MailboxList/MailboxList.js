import React, { Component } from 'react';
import { NavLink } from 'react-router-dom';

import './MailboxList.css'

export class MailboxList extends Component {

    constructor(props) {
        super(props);

        this.renderMessage = this.renderMessage.bind(this);
    }

    componentDidUpdate() {
        const url = this.props.location.pathname.split('/');
        if(url.length > 2) {
            this.scrollMessageIntoView(url[3]);
        }
    }

    scrollMessageIntoView(messageId) {
        const element = document.getElementById(messageId);
        if( element ) {
            element.scrollIntoView();
        }
    }

    renderHeader() {
        const recipientOrSender = this.props.match.path === '/inbox' ?
            <span className="sender">Sender</span>:
            <span className="recipient">To</span>;

        return (
            <li className="mailbox-list-header">
                {recipientOrSender}
                <span className="subject">Subject</span>
                <span className="date-sent">Date</span>
            </li>
        );
    }

    renderMessages() {
        return this.props.messages.map(this.renderMessage);
    }

    renderMessage(message) {
        const recipientOrSender =  this.props.match.path === '/inbox' ?
            <span className="sender">{message.sender}</span> :
            <span className="recipient">{message.recipient}</span>;

        return (
            <li className="mailbox-list-item" key={message.id} id={message.id}>
                <NavLink to={this.props.match.path + "/view/" + message.id}
                         className="mailbox-item-link"
                         activeClassName="is-active">
                    {recipientOrSender}
                    <span className="subject">{message.subject}</span>
                    <span className="date-sent">{message.dateSent}</span>
                </NavLink>
            </li>
        );
    }

    render() {
        return (
            <ol className="mailbox-list">
                { this.renderHeader() }
                { this.renderMessages() }
            </ol>
        );
    }
}