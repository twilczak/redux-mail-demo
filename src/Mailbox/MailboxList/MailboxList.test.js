import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { MemoryRouter } from 'react-router-dom';
import { MailboxList } from './MailboxList';

describe( MailboxList, () => {
    const messages = [{id: '123', dateSent: '2017.11.06', sender: 'testSender', recipient: 'testUser', subject: 'testing'}];

    it('renders and matches the snapshot', () => {
        const component = renderer.create(
            <MemoryRouter>
                <MailboxList match={{path: '/inbox'}} messages={messages}/>
            </MemoryRouter>
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('/inbox', () => {
        const component = mount(
            <MemoryRouter>
                <MailboxList match={{path: '/inbox'}} messages={messages}/>
            </MemoryRouter>
        );

        it('renders the mailbox headers', () => {
            const headers = component.find('.mailbox-list-header');
            expect(headers.find('.sender').text()).toContain('Sender');
            expect(headers.find('.subject').text()).toContain('Subject');
            expect(headers.find('.date-sent').text()).toContain('Date');
        });

        it('renders message contents', () => {
            const messageItem = component.find('.mailbox-list-item');
            expect(messageItem.find('.sender').text()).toContain('testSender');
            expect(messageItem.find('.subject').text()).toContain('testing');
            expect(messageItem.find('.date-sent').text()).toContain('2017.11.06');
        });
    });

    describe('/outbox', () => {
        const component = mount(
            <MemoryRouter>
                <MailboxList match={{path: '/outbox'}} messages={messages}/>
            </MemoryRouter>
        );

        it('renders the mailbox headers', () => {
            const headers = component.find('.mailbox-list-header');
            expect(headers.find('.recipient').text()).toContain('To');
            expect(headers.find('.subject').text()).toContain('Subject');
            expect(headers.find('.date-sent').text()).toContain('Date');
        });

        it('renders message contents', () => {
            const messageItem = component.find('.mailbox-list-item');
            expect(messageItem.find('.recipient').text()).toContain('testUser');
            expect(messageItem.find('.subject').text()).toContain('testing');
            expect(messageItem.find('.date-sent').text()).toContain('2017.11.06');
        });
    });

});