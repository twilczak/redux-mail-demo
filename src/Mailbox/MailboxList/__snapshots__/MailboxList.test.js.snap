// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class MailboxList extends _react.Component {

    constructor(props) {
        super(props);

        this.renderMessage = this.renderMessage.bind(this);
    }

    componentDidUpdate() {
        const url = this.props.location.pathname.split('/');
        if (url.length > 2) {
            this.scrollMessageIntoView(url[3]);
        }
    }

    scrollMessageIntoView(messageId) {
        const element = document.getElementById(messageId);
        if (element) {
            element.scrollIntoView();
        }
    }

    renderHeader() {
        const recipientOrSender = this.props.match.path === '/inbox' ?
        _react2.default.createElement('span', { className: 'sender', __source: { fileName: _jsxFileName, lineNumber: 30 }, __self: this }, 'Sender') :
        _react2.default.createElement('span', { className: 'recipient', __source: { fileName: _jsxFileName, lineNumber: 31 }, __self: this }, 'To');

        return (
            _react2.default.createElement('li', { className: 'mailbox-list-header', __source: { fileName: _jsxFileName, lineNumber: 34 }, __self: this },
                recipientOrSender,
                _react2.default.createElement('span', { className: 'subject', __source: { fileName: _jsxFileName, lineNumber: 36 }, __self: this }, 'Subject'),
                _react2.default.createElement('span', { className: 'date-sent', __source: { fileName: _jsxFileName, lineNumber: 37 }, __self: this }, 'Date')));


    }

    renderMessages() {
        return this.props.messages.map(this.renderMessage);
    }

    renderMessage(message) {
        const recipientOrSender = this.props.match.path === '/inbox' ?
        _react2.default.createElement('span', { className: 'sender', __source: { fileName: _jsxFileName, lineNumber: 48 }, __self: this }, message.sender) :
        _react2.default.createElement('span', { className: 'recipient', __source: { fileName: _jsxFileName, lineNumber: 49 }, __self: this }, message.recipient);

        return (
            _react2.default.createElement('li', { className: 'mailbox-list-item', key: message.id, id: message.id, __source: { fileName: _jsxFileName, lineNumber: 52 }, __self: this },
                _react2.default.createElement(_reactRouterDom.NavLink, { to: this.props.match.path + "/view/" + message.id,
                        className: 'mailbox-item-link',
                        activeClassName: 'is-active', __source: { fileName: _jsxFileName, lineNumber: 53 }, __self: this },
                    recipientOrSender,
                    _react2.default.createElement('span', { className: 'subject', __source: { fileName: _jsxFileName, lineNumber: 57 }, __self: this }, message.subject),
                    _react2.default.createElement('span', { className: 'date-sent', __source: { fileName: _jsxFileName, lineNumber: 58 }, __self: this }, message.dateSent))));



    }

    render() {
        return (
            _react2.default.createElement('ol', { className: 'mailbox-list', __source: { fileName: _jsxFileName, lineNumber: 66 }, __self: this },
                this.renderHeader(),
                this.renderMessages()));


    }} renders and matches the snapshot 1`] = `
<ol
  className="mailbox-list"
>
  <li
    className="mailbox-list-header"
  >
    <span
      className="sender"
    >
      Sender
    </span>
    <span
      className="subject"
    >
      Subject
    </span>
    <span
      className="date-sent"
    >
      Date
    </span>
  </li>
  <li
    className="mailbox-list-item"
    id="123"
  >
    <a
      aria-current={false}
      className="mailbox-item-link"
      href="/inbox/view/123"
      onClick={[Function]}
      style={undefined}
    >
      <span
        className="sender"
      >
        testSender
      </span>
      <span
        className="subject"
      >
        testing
      </span>
      <span
        className="date-sent"
      >
        2017.11.06
      </span>
    </a>
  </li>
</ol>
`;
